### PR TITLE
Initial draft of a Security Working Group governance

### PR DESCRIPTION
**Pull request by SidFaber**
This framework should outline the different roles & responsibilities within the Security Working Group.  The two primary jobs of the WG are to maintain subprojects and to handle all ROS vulnerabilities (both ROS1 and ROS2).

Most of the authority is vested in Approvers, they should be able to work independently and keep the overall working group informed during the monthly meetings.

### CODE REVIEW

File: README.md
Comment (kyrofa) [inline README.md]: I don't find this section particularly useful, I say we drop it until it makes sense to add.

File: README.md
Comment (kyrofa) [inline README.md]: I think this should be split out. I like the tooling WG's use of issue templates for adding new projects.

File: README.md
Comment (kyrofa) [inline README.md]: I think it's a little early for this, yet. We still need to come up with the disclosure policy, then we need to have a conversation with OR about our internal process. Let's remove this section for now to reflect the current status.

File: README.md
Comment (kyrofa) [inline README.md]: It's actually the last Tuesday of each month, but we've had to reschedule it each time as that had been a holiday the last few months :stuck_out_tongue: .

File: README.md
Comment (kyrofa) [inline README.md]: Need to add: "Meetings are recorded and available [on YouTube](https://www.youtube.com/playlist?list=PLpUh4ScdBhSMaEekJ8xeAAGmWUgR9S1K_)"

File: README.md
Comment (kyrofa) [inline README.md]: I don't think this needs to be defined. These are just non-members.

File: README.md
Comment (kyrofa) [inline README.md]: This is another one that I don't think needs to be defined.

File: README.md
Comment (kyrofa) [inline README.md]: I think you need to be an admin to do that, so this responsibility will lie with the WG chair/group admin(s).

File: README.md
Comment (kyrofa) [inline README.md]: Anyone can create a pull request, but only approvers can MERGE them.

File: README.md
Comment (kyrofa) [inline README.md]: Similar to above, let's remove this for now.

File: README.md
Comment (kyrofa) [inline README.md]: Members have issue triage privs.

File: README.md
Comment (kyrofa) [inline README.md:35]: Again taking from the tooling WG, let's make becoming a member/changing roles happen via issue templates.

**Review by kyrofa — CHANGES_REQUESTED** [review]
Thanks for this, @SidFaber! I made a number of comments, but I think we're just about there.

File: README.md
Comment (artivis) [inline README.md]:
```suggestion
The Security Working Group's mission is to advocate for and implement security within ROS 2.  The working group negotiates a balance between the benefits of security and the enablement of technology, striving for a secure-by-default design which can be customized to suit a targeted security profile.
```
or
```suggestion
The Security Working Group's mission is to advocate for and implement security features and tools within ROS 2.  The working group negotiates a balance between the benefits of security and the enablement of technology, striving for a secure-by-default design which can be customized to suit a targeted security profile.
```

File: README.md
Comment (artivis) [inline README.md]:
```suggestion
This document outlines the governance of the ROS 2 Security Working Group.
Updates to this document will be handled in the same manner as other project updates governed by the Working Group.
```

File: README.md
Comment (artivis) [inline README.md]:
```suggestion
A significant portion of the Security Working Group's mission is achieved by maintaining ROS subprojects. 
The following subprojects are owned by this Working Group:
```

File: README.md
Comment (artivis) [inline README.md]: I'd dislike the idea of discarding rules because they are cumbersome. I mean isn't it part of the point of a rule?
How about,
```suggestion
This guidelines are designed exclusively to help the Working Group achieve its mission, but are subject to change should they require to be actualized.
```

File: README.md
Comment (artivis) [inline README.md:24]: I feel like a `Communication channel` section is missing. At least the Matrix channel should be mentioned somewhere nop?

File: README.md
Comment (artivis) [inline README.md]:
```suggestion
* __Member__:  A participant in Security Working Group discussions. Responsiblities include:
```

File: README.md
Comment (kyrofa) [reply README.md]: Added template.

File: README.md
Comment (kyrofa) [reply README.md:35]: Added template and blurb.

File: README.md
Comment (kyrofa) [reply README.md]: It's not about discarding them if people don't like them, it's about maintaining flexibility. We're starting with a pretty light touch, but if we learn that the existing rules don't work, we should change them.

File: README.md
Comment (kyrofa) [reply README.md:24]: Yes indeed, good catch. Added it.

File: README.md
Comment (kyrofa) [reply README.md]: I think "security features" covers pretty much everything, including tools, when you're talking at the level of ROS 2 as a whole.

File: README.md
Comment (kyrofa) [reply README.md]: Removed all double-spacing after periods. They don't render in md anyway.

File: README.md
Comment (kyrofa) [reply README.md]: Removed all double-spacing after periods as they don't render in md anyway.

**Review by kyrofa — APPROVED** [review]

File: README.md
Comment (mikaelarguedas) [reply README.md]: While the rules are not yet fleshed out, it seems that this is one of the main missions of the working group. I would be in favor of keeping this section with a note "TBD of "Subject to changes" (similar to the list of project) rather than not having it mentionned at all in the readme

File: README.md
Comment (mikaelarguedas) [inline README.md]: If meeting alternate between timezones, it may be challenging for members of the timezones covered only every other meeting to meet the "majority of meetings" criterium

File: README.md
Comment (mikaelarguedas) [inline README.md:42]: It appears that both members and reviewers (actually anyone) can comment and review a PR on Github.
So my understanding is that reviewers have a responsibility / obligation to review PRs while member do not *have* to do it.

What would the workflow be in this context?
Users open PR
 - member triages it and assigns a reviewer 
   - reviewer iterates with contributor, once satisfied it assigns an approver to the PR
        - the approver iterates with contributor and once satisfied merges the PR

**Review by mikaelarguedas — COMMENTED** [review]
This looks great!

Not sure how cumbersome it can be for users that the repo has such a generic name, e.g. as a user, if I want to be involved in multiple working groups, I have to rename forks by hand to avoid collision. Maybe adding a prefix to the repo name would help.

Linkin to an inline comment here for visibility: https://github.com/ros-security/community/pull/1#discussion_r357533838

File: README.md
Comment (vmayoral) [reply README.md]: +1 for this. As we rotate the time, it should be reasonable that only timelines within a somewhat professional time-range are accounted for.

AFAIK, this is somehow the policy that Open Robotics is using for the ROS 2 TSC.

File: README.md
Comment (kyrofa) [reply README.md:42]: You're right, anyone can make or review a PR, or triage an issue; this is about responsibilities in the team more than about capabilities. If you're a member, you're responsible for helping to triage. If you're a reviewer, you're also responsible for reviewing. If you're an approver, you're also responsible for agreeing for the WG to take on new responsibilities (whether that's by approving a PR or by accepting new projects). To be clear, these build on each other-- approvers should also be working to triage, for example.

The workflow can be what you outlined, but it could be as simple as:

- PR is opened
  - Approver reviews and iterates until satisfied, and approves

I recommend some sanity checks, like ensuring there are two approvals before merging (and I'd say reviewer's approval counts), but that's a little out of scope here, and different repos might do it slightly differently.

File: README.md
Comment (kyrofa) [reply README.md]: That's fair, we could adjust the rotation a bit. The TSC says you can miss up to three meetings in a row, I think. We could move to something like that. I do think it's important for members to be limited to folks who are coming to meetings consistently, otherwise it's difficult to work as a team. Remember that non-members can still participate; membership comes with responsibility, and I think it's best if we avoid relying on folks we can never talk to.

File: README.md
Comment (kyrofa) [reply README.md]: I actually disagree. The primary responsibility of this WG is to drive the development of security features and best practices in ROS 2, not handle vuln reports. I honestly don't think the security WG should be handling this stuff, or we won't be able to leave membership nearly as open as I think we want it. It should actually be a separate group of engineers, both from the WG and from OR, and the mailing list should be maintained by OR, not this WG as a whole.

Could that actually be a subgroup of this WG? Perhaps a different member class? Yeah perhaps so, but I don't see any reason that the folks handling vuln reports also need to participate in new security features, so there's a bit of a disconnect there.

Long story short, I think we need to talk to OR before we propose or assume anything here.

Comment (mikaelarguedas) [reply README.md]: Fair enough, I think there was some misunderstanding from my side when reading the PR description

> The two primary jobs of the WG are to maintain subprojects and to handle all ROS vulnerabilities (both ROS1 and ROS2).

I now see that this PR clearly states a different mission statement (ROS 2 focused and not mentioning vulnerabilities):

> The Security Working Group's mission is to advocate for and implement security features within ROS 2. The working group negotiates a balance between the benefits of security and the enablement of technology, striving for a secure-by-default design which can be customized to suit a targeted security profile.

In which case no need to include the paragraph about vulnerability handling in here :+1: (and fully ROS 2 focused)

> Long story short, I think we need to talk to OR before we propose or assume anything here.

Yeah it may be beneficial to clarify with OR and working group members what the stance is on security vulnerabilities before advertising this as it may impact both user expectations (as a user I may assume that the security group/team is in charge of security patches) and membership interest

File: README.md
Comment (mikaelarguedas) [reply README.md]: > The TSC says you can miss up to three meetings in a row, I think. We could move to something like that.

That seems fair. On a round robin rotation you would expect that 2 out of 3 meeting can be attended by members regardless of their timezone.

Comment (kyrofa) [reply README.md]: Alright, updated.

**Comment by kyrofa** [comment]
I think everything has been resolved here. I'm going to go ahead and merge this, but that doesn't mean we can't continue iterating on it.

**Pull request merged**
